### PR TITLE
Research with local documents

### PR DESCRIPTION
**Pull request by ElishaKay**
Fixes an error when running on mac:

`docker-compose up --build`

The error (both on this branch & master branch) is:

 Failed to fetch http://deb.debian.org/debian-security/pool/updates/main/f/firefox-esr/firefox-esr_115.3.1esr-1%7edeb11u1_arm64.deb  404  Not Found [IP: 199.232.82.132 80]
#0 2.622 E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?

### CODE REVIEW

**Comment by assafelovic** [comment]
Thanks @ElishaKay ! Did you see that there are many conflicts?

**Comment by ElishaKay** [comment]
> Thanks @ElishaKay ! Did you see that there are many conflicts?

@assafelovic I rebased on top of master & resolved the conflicts while being very careful not to lose any changes of the past 3 weeks.

Tests pass for the "Short & fast summary" feature with "the Internet" & "My Documents" sources.

For the Detailed report (5-minutes), it looks like I'm getting errors because I didn't set my Tavily API Key? Are these known errors? Including the errors below:

`duckduckgo_search.exceptions.RatelimitException: https://links.duckduckgo.com/d.js?q=Maximizing+Freelance+Opportunities&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-311126008340105067307310865650937258949&ex=-1 202 Ratelimit`

And:

`requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://api.tavily.com`

**Comment by ElishaKay** [comment]
Tests pass for:
 - 2 minute internet report
 - 2 minute docs report
 - 5 minute docs report
 
 For the 5-minute internet report, it wrote the intro but then fails to generate subtopics & ends gracefully.

 I also tested this 5-minute internet report feature on the master branch, and in that case, the server crashed on the line in functions.py:
`sub_queries = json.loads(response)`



The function generate_subtopics_prompt function looks problematic to me in both this branch & the master branch since it seems to leverage global variables instead of explicitly passing any params into the function which would be much safer & easier to reason about:

```
def generate_subtopics_prompt() -> str:
    return """
                Provided the main topic:
                
                {task}
                
                and research data:
                
                {data}
                
                - Construct a list of subtopics which indicate the headers of a report document to be generated on the task. 
                - These are a possible list of subtopics : {subtopics}.
                - There should NOT be any duplicate subtopics.
                - Limit the number of subtopics to a maximum of {max_subtopics}
                - Finally order the subtopics by their tasks, in a relevant and meaningful order which is presentable in a detailed report
                
                "IMPORTANT!":
                - Every subtopic MUST be relevant to the main topic and provided research data ONLY!
                
                {format_instructions}
            """
```

**Comment by arsaboo** [comment]
Wouldn't this require modifications to the docker-compose to provide the mount path?

**Comment by arsaboo** [comment]
Don't we also need to modify the docker-compose to provide the mount path?

**Comment by ElishaKay** [comment]
> document-reports
> Don't we also need to modify the docker-compose to provide the mount path?
@arsaboo better to include the documents within the repository itself.
Reason being: if we mount local folders into a docker container, it can lead to permission issues

**Comment by olafgeibig** [comment]
@ElishaKay @arsaboo I think the mounting a docker volume is a good idea. This is being used by many dockerized applications. Maybe it could be optional to keep the default simple. I guess very often production deployments wouldn't like to have their docs in git.

**Comment by ElishaKay** [comment]
2-minute internet and documents reports work.
5-minute reports need work :)

**Comment by ElishaKay** [comment]
> @ElishaKay @arsaboo I think the mounting a docker volume is a good idea. This is being used by many dockerized applications. Maybe it could be optional to keep the default simple. I guess very often production deployments wouldn't like to have their docs in git. Thanks for this work. Can't wait that it's merged.

**Comment by ElishaKay** [comment]
@assafelovic,morning sir.
 - All 6 potential combos work: - 2-min, 5-min & resource-reports for both internet & document.
  - Don't believe it should have affected the multi-agent feature since I believe that folder was added after @proy9714's [original branch verged off master.](https://github.com/assafelovic/gpt-researcher/pull/429)

And best of all.... zero conflicts with master.
Would be thrilled if we could put other features on hold so as not to create new conflicts 😅

image

@proy9714 feel free to leave feedback here - solid stuff!
Let's do great things gentlemen

**Comment by assafelovic** [comment]
Thank you @proy9714 and @ElishaKay for this super super demanded feature! I'll stop all PRs until we get this through (except for hot fixes)! I will dive into it on Wed and we should release it next Monday

**Comment by assafelovic** [comment]
@ElishaKay can you update the documentation here: https://github.com/assafelovic/gpt-researcher/blob/master/docs/docs/gpt-researcher/tailored-research.md

**Comment by ElishaKay** [comment]
> @ElishaKay can you update the documentation here: https://github.com/assafelovic/gpt-researcher/blob/master/docs/docs/gpt-researcher/tailored-research.md

@assafelovic, Done.
Documentation added both for the pip package and running on localhost:8000 - with links and teaser on Readme.md

**Comment by ElishaKay** [comment]
Open issues for consideration:
A) we have 6 report_types and I only tested the 3 report_types available in the UI. @proy9714 do you think it will support all 6 potential report_types available from the pip package? If not, let's clarify that limit in the Docs.

B) @assafelovic, if the multi_agent feature is leveraging the same GPTResearcher Class, we should re-test the multi-agent feature on this branch? @assafelovic probably best to pass that to you since you're best acquainted with it

**Comment by proy9714** [comment]
> Open issues for consideration: A) we have 6 report_types and I only tested the 3 report_types available in the UI. @proy9714 do you think it will support all 6 potential report_types available from the pip package? If not, let's clarify that limit in the Docs.
> 
> B) @assafelovic, if the multi_agent feature is leveraging the same GPTResearcher Class, we should re-test the multi-agent feature on this branch? @assafelovic probably best to pass that to you since you're best acquainted with it

Hey! Yeah I guess it should work for all the report types, but I have not tested extensively. Also I should mention that a new issue was logged where the PydanticOutputParser from the current version of langchain fails to work with gpt-4o. So detailed report generation halts. I have suggested using the gpt-4-turbo model for the time being.

Here are the issues:

1. Issue no. #442 
2. Langchain Issue : https://github.com/langchain-ai/langchain/discussions/21635

**Comment by ElishaKay** [comment]
> [proy9714](/proy9714)

@proy9714 , I invited you now to collaborate on my fork of the repo - feel free to do whatever you gotta do.
I trust you bro!

**Comment by ElishaKay** [comment]
Looks like multi_agent cli feature broke 🙈
Will need to have a closer look this weekend

Error: TypeError: GPTResearcher.__init__() got an unexpected keyword argument 'verbose'

**Comment by assafelovic** [comment]
@ElishaKay check out this commit: https://github.com/assafelovic/gpt-researcher/commit/a091fd5e65ba6de3a7fda0ba6e9e22a2b73c2c5c

**Comment by ElishaKay** [comment]
> @ElishaKay did you merge master into this branch? I've added a param to GPTResearcher called `verbose` that disabled console prints if false (for production purposes).

@assafelovic After merging master to this branch, I had the same bug.

I tested the multi_agents feature on master branch now & got the same bug.

Error: TypeError: GPTResearcher.__init__() got an unexpected keyword argument 'verbose'

My theory is that the pip package wasn't updated.
Whereas most of the GPTResearcher instances in this repo leverage the locally defined GPTResearcher class

`from gpt_researcher.master.agent import GPTResearcher`

In the case of the multi_agents folder, it looks like the pip package is the source of this class:

`from gpt_researcher import GPTResearcher`

I'm updating the Product Roadmap now - it looks like our CI/CD logic should be:

When trying to merge to master:
 - run tests
 - update pip package
 - update docker package

In the meantime, I've removed the verbose keyword from the multi_agents folder & it runs to completion

File: docker-compose.yml
Comment (assafelovic) [inline docker-compose.yml:9]: @ElishaKay are you sure we should keep this static here? Probably keep it null as default?

File: .gitignore
Comment (assafelovic) [inline .gitignore:20]: This shouldn't be in docs. Docs is intended for documentation files

**Review by assafelovic — APPROVED** [review]
This is HUGE!

**Review by assafelovic — APPROVED** [review]

**Pull request merged**
